### PR TITLE
fetchgitlab: add deepClone, fetchSubmodules, leaveDotGit arguments

### PR DESCRIPTION
**Pull request by SuperSandro2000**
I also made fetchFromGitLab more similar to fetchFromGitHub.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/fetchgitlab/default.nix [hunk 1]
```diff
@@ -1,22 +1,28 @@
...
 
-  escapedSlug = replaceStrings ["." "/"] ["%2E" "%2F"] slug;
-  escapedRev = replaceStrings ["+" "%" "/"] ["%2B" "%25" "%2F"] rev;
+  fetcherArgs = (if useFetchGit then {
+    inherit rev deepClone fetchSubmodules leaveDotGit;
+    url = "https://${domain}/${slug}.git";
+  } else {
+    inherit name;
+    url = "https://${domain}/api/v4/projects/${escapedSlug}/repository/archive.tar.gz?sha=${escapedRev}";
+    meta.homepage = "https://${domain}/${slug}/";
```
Comment (SuperSandro2000) [inline pkgs/build-support/fetchgitlab/default.nix:24]:
```suggestion
```

File: pkgs/build-support/fetchgitlab/default.nix [hunk 2]
```diff
@@ -1,22 +1,28 @@
...
+  });
 in
 
-fetchzip ({
-  inherit name;
-  url = "https://${domain}/api/v4/projects/${escapedSlug}/repository/archive.tar.gz?sha=${escapedRev}";
-  meta.homepage = "https://${domain}/${slug}/";
-} // removeAttrs args [ "domain" "owner" "group" "repo" "rev" ]) // { inherit rev; }
+fetcher fetcherArgs // removeAttrs args [ "domain" "owner" "group" "repo" "rev" ] // { inherit rev; }
```
Comment (SuperSandro2000) [inline pkgs/build-support/fetchgitlab/default.nix:28]:
```suggestion
fetcher fetcherArgs // { meta.homepage = "https://${domain}/${slug}/"; inherit rev; }
```

File: pkgs/build-support/fetchgitlab/default.nix [hunk 3]
```diff
@@ -1,22 +1,28 @@
...
-  escapedSlug = replaceStrings ["." "/"] ["%2E" "%2F"] slug;
-  escapedRev = replaceStrings ["+" "%" "/"] ["%2B" "%25" "%2F"] rev;
+  fetcherArgs = (if useFetchGit then {
+    inherit rev deepClone fetchSubmodules leaveDotGit;
+    url = "https://${domain}/${slug}.git";
+  } else {
+    inherit name;
+    url = "https://${domain}/api/v4/projects/${escapedSlug}/repository/archive.tar.gz?sha=${escapedRev}";
+    meta.homepage = "https://${domain}/${slug}/";
+  });
```
Comment (SuperSandro2000) [inline pkgs/build-support/fetchgitlab/default.nix:25]:
```suggestion
  }) // removeAttrs args [ "domain" "owner" "group" "repo" "rev" ] // { inherit name; };
```

File: pkgs/build-support/fetchgitlab/default.nix [hunk 4]
```diff
@@ -1,22 +1,27 @@
...
+  useFetchGit = deepClone || fetchSubmodules || leaveDotGit;
+  fetcher = if useFetchGit then fetchgit else fetchzip;
 
-  escapedSlug = replaceStrings ["." "/"] ["%2E" "%2F"] slug;
-  escapedRev = replaceStrings ["+" "%" "/"] ["%2B" "%25" "%2F"] rev;
+  fetcherArgs = (if useFetchGit then {
+    inherit rev deepClone fetchSubmodules leaveDotGit;
+    url = "https://${domain}/${slug}.git";
+  } else {
+    inherit name;
```
Comment (SuperSandro2000) [inline pkgs/build-support/fetchgitlab/default.nix:22]:
```suggestion
```

File: pkgs/build-support/fetchgitlab/default.nix [hunk 5]
```diff
@@ -1,22 +1,26 @@
...
 
-with lib;
-
 let
-  slug = concatStringsSep "/"
-    ((optional (group != null) group) ++ [ owner repo ]);
+  slug = lib.concatStringsSep "/" ((lib.optional (group != null) group) ++ [ owner repo ]);
+
+  escapedSlug = lib.replaceStrings [ "." "/" ] [ "%2E" "%2F" ] slug;
+  escapedRev = lib.replaceStrings [ "+" "%" "/" ] [ "%2B" "%25" "%2F" ] rev;
```
Comment (SuperSandro2000) [inline pkgs/build-support/fetchgitlab/default.nix:13]:
```suggestion
  slug = lib.concatStringsSep "/" ((lib.optional (group != null) group) ++ [ owner repo ]);
  escapedSlug = lib.replaceStrings [ "." "/" ] [ "%2E" "%2F" ] slug;
  escapedRev = lib.replaceStrings [ "+" "%" "/" ] [ "%2B" "%25" "%2F" ] rev;
  passthruAttrs = removeAttrs args [ "domain" "owner" "group" "repo" "rev" ];
```

File: pkgs/build-support/fetchgitlab/default.nix [hunk 6]
```diff
@@ -1,22 +1,26 @@
...
+  fetcher = if useFetchGit then fetchgit else fetchzip;
 
-  escapedSlug = replaceStrings ["." "/"] ["%2E" "%2F"] slug;
-  escapedRev = replaceStrings ["+" "%" "/"] ["%2B" "%25" "%2F"] rev;
+  fetcherArgs = (if useFetchGit then {
+    inherit rev deepClone fetchSubmodules leaveDotGit;
+    url = "https://${domain}/${slug}.git";
+  } else {
+    url = "https://${domain}/api/v4/projects/${escapedSlug}/repository/archive.tar.gz?sha=${escapedRev}";
+  }) // removeAttrs args [ "domain" "owner" "group" "repo" "rev" ] // { inherit name; };
```
Comment (SuperSandro2000) [inline pkgs/build-support/fetchgitlab/default.nix:23]:
```suggestion
  }) // passthruAttrs // { inherit name; };
```

**Comment by TethysSvensson** [comment]
LGTM, though I haven't tested it.

**Pull request merged**